### ISSUE TITLE
cross encoder

### ISSUE THREAD

**Issue by ejklike** [ejklike/cnmt1.0#4]
"""
Implementation of "Attention is All You Need"
"""

import torch.nn as nn

from onmt.encoders.encoder import EncoderBase
from onmt.modules import MultiHeadedAttention
from onmt.modules.position_ffn import PositionwiseFeedForward
from onmt.utils.misc import sequence_mask

class TransformerCrossEncoderLayer(nn.Module):
    """
    A single layer of the transformer encoder.

    Args:
        d_model (int): the dimension of keys/values/queries in
                   MultiHeadedAttention, also the input size of
                   the first-layer of the PositionwiseFeedForward.
        heads (int): the number of head for MultiHeadedAttention.
        d_ff (int): the second-layer of the PositionwiseFeedForward.
        dropout (float): dropout probability(0-1.0).
    """

    def __init__(self, d_model, heads, d_ff, dropout, attention_dropout,
                 max_relative_positions=0):
        super(TransformerCrossEncoderLayer, self).__init__()
        self.self_attn = MultiHeadedAttention(
            heads, d_model, dropout=dropout,
            max_relative_positions=max_relative_positions)
        self.cross_attn = MultiHeadedAttention(
            heads, d_model, dropout=attention_dropout)
        self.feed_forward = PositionwiseFeedForward(d_model, d_ff, dropout)
        self.layer_norm_1 = nn.LayerNorm(d_model, eps=1e-6)
        self.layer_norm_2 = nn.LayerNorm(d_model, eps=1e-6)
        self.drop = nn.Dropout(dropout)

    def forward(self, x, x_mask, y, y_mask):
        """
        Args:
            inputs (FloatTensor): ``(batch_size, src_len, model_dim)``
            mask (LongTensor): ``(batch_size, 1, src_len)``

        Returns:
            (FloatTensor):

            * outputs ``(batch_size, src_len, model_dim)``
        """
        # Self-attn
        x_norm = self.layer_norm_1(x)
        x_query, _ = self.self_attn(x_norm, x_norm, x_norm,
                                    mask=x_mask, attn_type="self")
        x_query = self.drop(x_query) + x

        # Cross-attn
        x_query_norm = self.layer_norm_2(x_query)
        # key, value, query
        # memory_bank, memory_bank, query_norm,
        mid, _ = self.cross_attn(y, y, x_query_norm,
                                 mask=y_mask, attn_type="self")
        out = self.drop(mid) + x_query

        return self.feed_forward(out)

    def update_dropout(self, dropout, attention_dropout):
        self.self_attn.update_dropout(attention_dropout)
        self.cross_attn.update_dropout(attention_dropout)
        self.feed_forward.update_dropout(dropout)
        self.drop.p = dropout

class TransformerCrossEncoder(EncoderBase):
    """The Transformer encoder from "Attention is All You Need"
    :cite:`DBLP:journals/corr/VaswaniSPUJGKP17`

    .. mermaid::

       graph BT
          A[input]
          B[multi-head self-attn]
          C[feed forward]
          O[output]
          A --> B
          B --> C
          C --> O

    Args:
        num_layers (int): number of encoder layers
        d_model (int): size of the model
        heads (int): number of heads
        d_ff (int): size of the inner FF layer
        dropout (float): dropout parameters
        embeddings (onmt.modules.Embeddings):
          embeddings to use, should have positional encodings

    Returns:
        (torch.FloatTensor, torch.FloatTensor):

        * embeddings ``(src_len, batch_size, model_dim)``
        * memory_bank ``(src_len, batch_size, model_dim)``
    """

    def __init__(self, num_layers, d_model, heads, d_ff, dropout,
                 attention_dropout, embeddings, max_relative_positions,
                 encoder):
        super(TransformerCrossEncoder, self).__init__()

        self.embeddings = embeddings
        self.encoder = encoder
        self.transformer = nn.ModuleList(
            [TransformerCrossEncoderLayer(
                d_model, heads, d_ff, dropout, attention_dropout,
                max_relative_positions=max_relative_positions)
             for i in range(num_layers)])
        self.layer_norm = nn.LayerNorm(d_model, eps=1e-6)

    @classmethod
    def from_opt(cls, opt, embeddings, encoder):
        """Alternate constructor."""
        return cls(
            opt.enc_layers,
            opt.enc_rnn_size,
            opt.heads,
            opt.transformer_ff,
            opt.dropout[0] if type(opt.dropout) is list else opt.dropout,
            opt.attention_dropout[0] if type(opt.attention_dropout)
            is list else opt.attention_dropout,
            embeddings,
            opt.max_relative_positions,
            encoder)

    def forward(self, src, tgt):
        """See :func:`EncoderBase.forward()`"""

        pad_idx = self.embeddings.word_padding_idx
        src_words = src[:, :, 0].transpose(0, 1) # [batch_size, word_len]
        src_mask = src_words.data.eq(pad_idx).unsqueeze(1)

        tgt_words = tgt[:, :, 0].transpose(0, 1) # [batch_size, word_len]
        tgt_mask = tgt_words.data.eq(pad_idx).unsqueeze(1)

        emb = self.embeddings(src)
        out = emb.transpose(0, 1).contiguous()

        _, tgt, _ = self.encoder(tgt)
        tgt = tgt.transpose(0, 1).contiguous()

        # Run the forward pass of every layer of the tranformer.
        for layer in self.transformer:
            out = layer(out, src_mask, tgt, tgt_mask)
        out = self.layer_norm(out)

        return out.transpose(0, 1).contiguous()

    def update_dropout(self, dropout, attention_dropout):
        self.embeddings.update_dropout(dropout)
        self.encoder.update_dropout(dropout)
        for layer in self.transformer:
            layer.update_dropout(dropout, attention_dropout)
